src/tester.rs

```diff
@@ -569,30 +583,103 @@ fn run_cmd(
     test_fname: &str,
     mut cmd: Command,
 ) -> (ExitStatus, Vec<u8>, Vec<u8>) {
+    // The basic sequence here is:
+    //   1) Spawn the command
+    //   2) Read everything from stderr & stdout until they are both disconnected
+    //   3) wait() for the command to finish
+
     let mut child = cmd
         .stderr(process::Stdio::piped())
         .stdout(process::Stdio::piped())
         .spawn()
         .unwrap_or_else(|_| fatal(&format!("Couldn't run command {:?}.", cmd)));
 
-    let mut looped = 1;
+    let mut stderr = FileDescriptor::dup(child.stderr.as_ref().unwrap()).unwrap();
+    let mut stdout = FileDescriptor::dup(child.stdout.as_ref().unwrap()).unwrap();
+
+    let mut cap_stderr = Vec::new();
+    let mut cap_stdout = Vec::new();
+    let mut pollfds = [
+        pollfd {
+            fd: FileDescriptor::dup(&stderr)
+                .unwrap()
+                .into_socket_descriptor(),
+            events: POLLERR | POLLIN | POLLHUP,
+            revents: 0,
+        },
+        pollfd {
+            fd: FileDescriptor::dup(&stdout)
+                .unwrap()
+                .into_socket_descriptor(),
+            events: POLLERR | POLLIN | POLLHUP,
+            revents: 0,
+        },
+    ];
+    let mut buf = [0; READBUF];
+    let start = Instant::now();
+    let mut last_warning = Instant::now();
+    let mut next_warning = last_warning
+        .checked_add(Duration::from_secs(TIMEOUT))
+        .unwrap();
     loop {
-        match child.wait_timeout(Duration::from_secs(TIMEOUT)).unwrap() {
-            Some(_) => break,
-            None => {
-                if inner.test_threads == 1 {
-                    eprint!("running for over {} seconds... ", TIMEOUT * looped);
-                } else {
-                    eprintln!(
-                        "\nlang_tests::{} ... has been running for over {} seconds",
-                        test_fname,
-                        TIMEOUT * looped
-                    );
-                }
+        let timeout = {
+            // FIXME: When Rust 1.39.0 is out we can replace this mess with duration_since_checked.
+            let t = Instant::now();
+            if t > next_warning {
+                Duration::from_secs(1)
+            } else {
+                next_warning.duration_since(t)
             }
         };
-        looped += 1;
+        if let Ok(_) = poll(&mut pollfds, Some(timeout)) {
+            if pollfds[0].revents & POLLIN == POLLIN {
```

Comment:
It looks like we only check the first file descriptor in `pollfds`, which is `stderr`, when does  `stdout` get checked?

Reply:
Line 649(ish) handles `stdout`.

Reply:
Ah yes, not sure why I missed this!